### ISSUE TITLE
Add support for frozen output layer

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#4760]
Moved from: https://github.com/deeplearning4j/nd4j/issues/2706 - cc @ionspin

#### Issue Description
> As discussed on gitter, currently frozen layers don't support output layers. If a frozen layer wraps an output layer an exception is thrown.
```
        if (insideLayer instanceof OutputLayer) {
            throw new IllegalArgumentException("Output Layers are not allowed to be frozen " + layerId());
        }
```

#### Version Information

* Deeplearning4j 0.9.2-SNAPSHOT

#### Contributing

I'd like to help out with this issue, but would definitely need some guidance

**Comment by AlexDBlack** [comment]
OK, so the starting point here would be to simply remove that exception and add some unit tests - I'm not sure if it'll work fine or not TBH. The exception was more of a design decision - at the time we didn't think of a legitimate use case that involved freezing the output layer. In most cases it still doesn't make sense, but there are exceptions.

**Comment by raver119** [comment]
I guess it's safe to close this issue now?

**Comment by AlexDBlack** [comment]
No. It *might* work currently, but there's definitely no unit tests for this.

**Comment by ionspin** [comment]
Sorry for my delay on this, I have been moving between cities, so I didn't have much free time. I'll work on it this weekend.

**Comment by ionspin** [comment]
I have tried to understand the the flow and I think this is the current state:
* Currently when there is a frozen layer in a multi-layer network model, all of the layers that came before it will not be trained, as FrozenLayer will return default gradient and null as epsilon.
* Also there is this code in calcBackpropGradients method of MultiLayerNetwork.java https://github.com/deeplearning4j/deeplearning4j/blob/355747d1688e664e80fc3f412fc7a1268af2d958/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/multilayer/MultiLayerNetwork.java#L1541-L1544
that also points to backprop being stopped at first frozen layer.
* On the other hand when I was playing around with MNIST GAN, it was suggested that using the Sgd updater with learning rate set to 0.0 would suffice, as I needed to freeze last three and the output layer, and train the layers leading up to those.

So with this intro, would it make more sense to leave the current FrozenLayer implementation as is, and introduce a new BackpropEnabledFrozenLayer (or similar) class that would use Sgd to freeze the parameters but allow the backprop to continue?

**Comment by ionspin** [comment]
@AlexDBlack I'll go ahead and spend some time on making the necessary changes to enable frozen layers to return gradients, but I just need your guidance on what would be a more suitable solution:
* Modify the current FrozenLayer implementation by adding a configuration parameter that would enable backpropagation (with default being false for backwards compatibility and performance reasons)
* Create a new implementation of a layer that would allow backpropagation and leave the FrozenLayer  as it is
* Something else entirely

**Comment by AlexDBlack** [comment]
Supported with tests here: https://github.com/deeplearning4j/deeplearning4j/pull/5106
